### PR TITLE
Retrieve function source description from comments

### PR DESCRIPTION
**Pull request by sharkAndshark**
If a PostgreSQL function has an SQL comment, it will try to parse as JSON and use its values to override the auto-generated TileJSON.  It is recommended to use this form when creating comments to ensure valid JSON values.

```sql
DO $do$ BEGIN
    EXECUTE 'COMMENT ON FUNCTION YOUR_FUNCTION (ARG1_TYPE,ARG2_TYPE,..ARGN_TYPE) IS $tj$' || $$
    {
      "description": "description override",
      ...
    }
    $$::json || '$tj$';
END $do$;
```

Partially implements #822

### CODE REVIEW

File: martin/src/pg/function_source.rs [hunk 1]
```diff
@@ -13,6 +15,14 @@ use crate::pg::pool::PgPool;
 use crate::pg::PgError::PostgresError;
 use crate::pg::Result;
 
+#[derive(Serialize, Deserialize, Debug)]
```
Comment (nyurik) [inline martin/src/pg/function_source.rs]: use `TileJSON` object here - no point in creating a new struct.  Simply parse `TileJSON` from the comment

File: martin/src/pg/scripts/query_available_function.sql [hunk 2]
```diff
@@ -29,17 +29,28 @@ WITH inputs AS (
                 jsonb_agg(parameter_name::text ORDER BY ordinal_position) as out_names
          FROM information_schema.parameters
          WHERE parameter_mode = 'OUT'
-         GROUP BY specific_name)
+         GROUP BY specific_name),
+    comments AS (
+        SELECT
+	        n.nspname AS schema,
+	        P.proname AS name,
+	        obj_description ( P.oid ) AS description
```
Comment (nyurik) [inline martin/src/pg/scripts/query_available_function.sql]:
```suggestion
	        obj_description(P.oid, 'pg_proc') AS description
```

File: tests/fixtures/functions/function_Mixed_Name.sql [hunk 3]
```diff
@@ -13,3 +13,25 @@ RETURNS TABLE("mVt" bytea, key text) AS $$
       WHERE "Geom" && ST_Transform(ST_TileEnvelope("Z", x, y), 4326)
   ) as tile WHERE geom IS NOT NULL) src
 $$ LANGUAGE SQL IMMUTABLE STRICT PARALLEL SAFE;
+
+COMMENT ON FUNCTION "MixedCase"."function_Mixed_Name" (int4,INT4,INT4) IS $tilejson$
```
Comment (nyurik) [inline tests/fixtures/functions/function_Mixed_Name.sql]: See if you can modify these somehow to auto-verified with `'{...}'::json` approach instead of a unverified string

Comment (sharkAndshark) [reply tests/fixtures/functions/function_Mixed_Name.sql]: Weired. PostgreSQL complained syntax error. Need a little bit research on.
```SQL
COMMENT ON FUNCTION osm_trees(INTEGER,INTEGER,INTEGER) IS ('{"a":1}'::json::TEXT)
```

Comment (sharkAndshark) [reply tests/fixtures/functions/function_Mixed_Name.sql]: Seems it only receive literal value?  [doc](https://www.postgresql.org/docs/15/sql-comment.html)

Comment (nyurik) [reply tests/fixtures/functions/function_Mixed_Name.sql]: @sharkAndshark use this method:

```sql
DO $do$ BEGIN
    EXECUTE 'COMMENT ON FUNCTION public.function_null (INT4,INT4,INT4) IS $tj$' || $$
    {
      "minzoom": 0,
      "maxzoom": 18,
      "bounds": [-180, -85, 180, 85],
      "vector_layers": []
    }
    $$::json || '$tj$';
END $do$;
```

**Comment by sharkAndshark** [comment]
Hi @nyurik  Could you give me some suggestions for fixing the errors in SQL files? I've tried many times.
It's weired. The `setup-database` step in CI failed.

Update:
Seems it's due to the case sentive issue.But I think I shouldn't double quoting every sql file.Fix it and do the method you suggested later

**Comment by nyurik** [comment]
> Hi @nyurik Could you give me some suggestions for fixing the errors in SQL files? I've tried many times. It's weired. The `setup-database` step in CI failed.
> 
> Update: Seems it's due to the case sentive issue.But I think I shouldn't double quoting every sql file.Fix it and do the method you suggested later

**usually** you don't need to quote any identifiers (namespaces, tables, functions and their arguments, etc) in SQL, as long as they don't use any special characters or reserved keywords. BUT, in some very rare cases, someone may have those weird cases. So in our testing, we must make sure Martin supports it properly. And for that, I have a few tests like `"MixedCase"` namespaces, as well as functions, etc.

Note that the generated SQL automatically always uses quotes identifiers to avoid ambiguity.

**Comment by nyurik** [comment]
I reworked your patch in https://github.com/sharkAndshark/martin/pull/50 -- please take a look and merge if agree.

Basic idea:
* tilejson from a comment is stored as a simple json value
* when creating the final tilejson, it uses generated tilejson as the "base", and applies all fields from the comment's tilejson as a "patch" (override) using [json-patch](https://crates.io/crates/json-patch) crate

P.S. I don't think you should add tilejson to ALL unit tests. Instead, add it to just a few, with some variation of which fields are added, and possibly even deleted (set them to null).

**Comment by sharkAndshark** [comment]
Agreed.I realized the tests part and I'm in a conference,merge it later.

**Comment by sharkAndshark** [comment]
To achieve isolation, it is recommended to have one unit test for each observable behavior. Therefore, it might be necessary to remove some comments from the function. However, there is a concern that this level of granularity might be excessive?(And we can not see them on `debug.html` then) @nyurik I'm looking forward your guide.

**Comment by nyurik** [comment]
@sharkAndshark lets keep things simple :)  I think we need to cover these cases. And I think you can do this with just two function comments.

### Test 1
* set on `"MixedCase"."function_Mixed_Name"` - to test comments work on weird named functions
* override a single field:  set the `description` in a function comment -- this should override the whatever description is set from the full ID automatically
* add a new field:  set `vector_layers` field with an array of whatever stuff the function is producing

### Test 2
* set comment on `public.function_zxy_query` -- normal name, different param combination
* delet a field - in another function comment test, set `description` to `null`. No need to set any other fields at all
* add some field that is NOT part of tilejson, e.g. `"myvalue": { "foo": "bar" }`  (or whatever similar). This will test that we can add arbitrary complex content

Just doing these two tests, you should be good enough.

**Comment by sharkAndshark** [comment]
I hope the doc is not too bad as English is not my first language and I'm still learning it.

File: tests/fixtures/functions/function_zxy_query.sql [hunk 4]
```diff
@@ -16,3 +16,14 @@ BEGIN
   RETURN mvt;
 END
 $$ LANGUAGE plpgsql IMMUTABLE STRICT PARALLEL SAFE;
+
+DO $do$ BEGIN
+    EXECUTE 'COMMENT ON FUNCTION public.function_zxy_query (INT4, INT4, INT4, JSON) IS $tj$' || $$
+    {
+      "tilejson": "3.0.0",
```
Comment (nyurik) [inline tests/fixtures/functions/function_zxy_query.sql]: Don't set the version or tiles fields here and above

Comment (sharkAndshark) [reply tests/fixtures/functions/function_zxy_query.sql:23]: I tried before, It's required for `tilejson` crate.

File: martin/src/pg/function_source.rs [hunk 5]
```diff
@@ -33,6 +34,18 @@ pub async fn query_available_function(pool: &PgPool) -> Result<SqlFuncInfoMapMap
             let output_record_names = jsonb_to_vec(&row.get("output_record_names"));
             let input_types = jsonb_to_vec(&row.get("input_types")).expect("Can't get input types");
             let input_names = jsonb_to_vec(&row.get("input_names")).expect("Can't get input names");
+            let tilejson = if let Some(text) = row.get("description") {
+                match serde_json::from_str::<Value>(text) {
```
Comment (nyurik) [inline martin/src/pg/function_source.rs:38]: parsing now works because it now uses `from_str::<Value>` instead of `from_str::<TileJSON>`.  Value allows any valid json.

**Review by nyurik — APPROVED** [review]
great work, thanks!

File: martin/src/pg/function_source.rs (hunk 5 above)
Comment (sharkAndshark) [reply martin/src/pg/function_source.rs:38]: Yeah. hehe. I realized that lol. Cool. ✅

**Comment by nyurik** [comment]
Thanks @sharkAndshark , i may do another round of docs cleanup later, but this is good to merge for now.  Do you perhaps want to do the same thing for tables?  I suspect it is possible to add comments to them too.

**Comment by sharkAndshark** [comment]
Yeah sure. I'd love to do that. A issue ticket for discussion first maybe? @nyurik

**Comment by nyurik** [comment]
technically that's the same issue #822 issue - although it is more of an umbrella issue to cover all tilejson modifications by the user.  Eventually I think we should just add an extra field to the yaml config, and that field will work the same way as the function comment
